### ISSUE TITLE
Nomad backend support?

### ISSUE THREAD

**Issue by JalfResi** [ContainerSSH/ContainerSSH#384]
## Please describe what you would like to see in ContainerSSH

I've been looking for something exactly like containerSSH but my cluster backend is managed by Nomad. Is there support for Nomad planned on the horizon?

## Please describe the solution you'd like
I'd love to be able to direct ssh users to their own container, whose lifecycle is managed via nomad (much like the kubernetes backend)

## Please describe your use case
Providing a terminal based game over ssh, I'd like to isolate players from each other using containers (scratch container with only my game binary within it). However, my backend services are currently managed via Nomad, and I'd like to continue to use Nomad and supporting tools to manage the docker containers running within my cluster, including each users container that is spun up on demand

**Comment by janosdebugs** [comment]
Hey @JalfResi,

So far we haven't looked at Nomad as a possible backend. Looking at the [API docs](https://www.nomadproject.io/api-docs/) it seems that it is missing several important features for ContainerSSH, feel free to correct please:

- The ability to launch individual containers (could be mitigated by using deployments).
- The ability to attach to a container using interactive terminals (it seems you can only fetch logs, not supply STDIN or request an interactive terminal).
- The ability to launch additional tasks dynamically within an existing container (required for the ContainerSSH agent to work correctly).

I'm certain that eventually support for Nomad could be added, but it does not seem to be as straight forward as the Docker and Kubernetes. We are currently focussed on the authentication side of things, so I don't think we can add Nomad in a reasonable timeline. :(

**Comment by RickyGrassmuck** [comment]
A bunch has changed with nomad over the last 2 years and all of the functionality you have listed are supported.

For the first item, I think the best approach would be to make use of parameterized jobs -https://developer.hashicorp.com/nomad/api-docs/jobs#dispatch-job

Exec'ing into a container can be done via the API as well - https://developer.hashicorp.com/nomad/api-docs/allocations#exec-allocation

The last one can be accomplished using the Task API provided by Nomad that allows mounting a unix socket in a running job which it can use to communicate with Nomad - https://developer.hashicorp.com/nomad/api-docs/task-api

Nomad recently added support for Workload identities as well which may be of interest for Auth purposes too.

**Comment by tsipinakis** [comment]
I agree that hashicorp is in an uncertain state right now so I would hold off on merging for a while until things are stabilized. However it does present an interesting target.

> It would also require significant work on the ContainerSSH agent to be able to spawn off full TTY shells.

This would not be a downside IMO, the opposite it's _very_ useful for current backends as well, as we can decouble the shell from using the kube API, after it's launched saving resources and making the connections to the container more robust by being direct point to point

**Comment by tsipinakis** [comment]
Indeed, but no reason why both ways of spawning a shell cannot be supported, kube api OR agent-shell.
